src/Publisher.php

```diff
@@ -0,0 +1,151 @@
+     * This must be true if you are using this with the "subsites" module.
+     * Please note that this form of caching requires all URLs to be provided absolute
+     * (not relative to the webroot) via {@link SiteTree->AbsoluteLink()}.
+     */
+    private static $domain_based_caching = false;
+
+    /**
+     * @param string $url
+     * @return HTTPResponse
+     */
+    public function generatePageResponse($url)
+    {
+        if (Director::is_relative_url($url)) {
+            $url = Director::absoluteURL($url);
+        }
+        $urlParts = parse_url($url);
+        if (!empty($urlParts['query'])) {
+            parse_str($urlParts['query'], $getVars);
+        } else {
+            $getVars = [];
+        }
+        // back up requirements backend
+        $origRequirements = Requirements::backend();
+        Requirements::set_backend(Requirements_Backend::create());
+
+        $origThemes = SSViewer::get_themes();
+        $staticThemes = self::config()->get('static_publisher_themes');
+        if ($staticThemes) {
+            SSViewer::set_themes($staticThemes);
+        } else {
+            // get the themes raw from config to prevent the "running from the CMS" problem where no themes are live
+            $rawThemes = SSViewer::config()->uninherited('themes');
+            SSViewer::set_themes($rawThemes);
+        }
+        try {
+            // try to add all the server vars that would be needed to create a static cache
+            $request = HTTPRequestBuilder::createFromVariables(
+                [
+                '_SERVER' => [
+                    'REQUEST_URI' => isset($urlParts['path']) ? $urlParts['path'] : '',
+                    'REQUEST_METHOD' => 'GET',
+                    'REMOTE_ADDR' => '127.0.0.1',
+                    'HTTPS' => $urlParts['scheme'] == 'https' ? 'on' : 'off',
+                    'QUERY_STRING' => isset($urlParts['query']) ? $urlParts['query'] : '',
+                    'REQUEST_TIME' => DBDatetime::now()->getTimestamp(),
+                    'REQUEST_TIME_FLOAT' => (float) DBDatetime::now()->getTimestamp(),
+                    'HTTP_HOST' => $urlParts['host'],
+                    'HTTP_USER_AGENT' => 'silverstripe/staticpublishqueue',
+                ],
+                '_GET' => $getVars,
+                '_POST' => [],
+                ],
+                ''
+            );
+            $app = $this->getHTTPApplication();
+            $response = $app->handle($request);
+        } catch (HTTPResponse_Exception $e) {
+            $response = $e->getResponse();
+        } finally {
+            // restore backends
+            SSViewer::set_themes($origThemes);
+            Requirements::set_backend($origRequirements);
+            DataObject::singleton()->flushCache();
+        }
+        return $response;
+    }
+
+    /**
+     * @return HTTPApplication
+     */
+    protected function getHTTPApplication()
+    {
+        $kernel = new CoreKernel(BASE_PATH);
```

Comment:
Probably better to use `Injector::inst()->create(Kernel::class, BASE_PATH)` unless there's a reason to skip the injector

Reply:
Kernels can't be injected because they are created before the Manifests / Injector is ready (the Kernels create the injector instances) - so whilst we could use the injector, it wouldn't make too much sense, it'd be better for someone to create their own Publisher and override the `getHTTPApplication` method IMO